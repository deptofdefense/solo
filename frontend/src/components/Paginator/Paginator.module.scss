.pageLink {
  border-radius: 30% !important;
  text-decoration: none !important;
  margin: 0 1rem !important;
}

.disabled {
  color: lightgray !important;
  cursor: default !important;
}
