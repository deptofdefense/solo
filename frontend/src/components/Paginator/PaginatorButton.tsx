import React from "react";
import classNames from "classnames";
import { Button } from "solo-uswds";
import classes from "./Paginator.module.scss";

interface PaginatorButtonProps {
  disabled?: boolean;
  isPageNumber?: boolean;
  active?: boolean;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
}

const PaginatorButton: React.FC<PaginatorButtonProps> = ({
  disabled = false,
  isPageNumber = false,
  active = false,
  children,
  ...rest
}) => {
  return (
    <Button
      className={classNames(classes.pageLink, {
        [classes.disabled]: disabled
      })}
      unstyled={!isPageNumber || !active}
      outline={isPageNumber && active}
      onClick={!disabled ? onclick : undefined}
      {...rest}
    >
      {children}
    </Button>
  );
};

export default PaginatorButton;
