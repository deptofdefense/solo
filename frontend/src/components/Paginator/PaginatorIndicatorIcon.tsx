import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faChevronLeft,
  faChevronRight
} from "@fortawesome/free-solid-svg-icons";

interface IndicatorProps {
  left?: boolean;
  pad?: boolean;
}

const PaginatorIndicatorIcon: React.FC<IndicatorProps> = ({
  left = false,
  pad = false
}) => (
  <FontAwesomeIcon
    size="sm"
    icon={left ? faChevronLeft : faChevronRight}
    className={pad ? (left ? "margin-right-1" : "margin-left-1") : ""}
  />
);

export default PaginatorIndicatorIcon;
