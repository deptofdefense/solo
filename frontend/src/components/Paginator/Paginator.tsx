import React, { PropsWithChildren } from "react";
import { TableInstance } from "react-table";
import PaginatorIndicator from "./PaginatorIndicatorIcon";
import PaginatorButton from "./PaginatorButton";

interface PaginatorProps<T extends object> {
  table: TableInstance<T>;
}
const Paginator = <T extends object>({
  table: {
    canPreviousPage,
    canNextPage,
    previousPage,
    nextPage,
    pageCount,
    gotoPage,
    state: { pageIndex: idx }
  }
}: PropsWithChildren<PaginatorProps<T>>) => {
  const startIdx = idx >= 2 ? idx - 2 : 0;
  const endIdx = idx <= pageCount - 3 ? idx + 3 : pageCount;
  const listSize = endIdx - startIdx;

  return (
    <div className="grid-row flex-justify-center margin-top-3">
      <PaginatorButton onClick={() => gotoPage(0)} disabled={!canPreviousPage}>
        <PaginatorIndicator left />
        <PaginatorIndicator left pad />
        First
      </PaginatorButton>

      <PaginatorButton
        onClick={() => previousPage()}
        disabled={!canPreviousPage}
      >
        <PaginatorIndicator left pad />
        Previous
      </PaginatorButton>

      {Array.from({ length: listSize }).map((_, i) => {
        const pageNum = i + startIdx;
        return (
          <PaginatorButton
            isPageNumber
            active={pageNum === idx}
            key={pageNum}
            onClick={() => gotoPage(pageNum)}
          >
            {pageNum + 1}
          </PaginatorButton>
        );
      })}

      <PaginatorButton onClick={nextPage} disabled={!canNextPage}>
        Next
        <PaginatorIndicator pad />
      </PaginatorButton>

      <PaginatorButton
        onClick={() => gotoPage(pageCount - 1)}
        disabled={!canNextPage}
      >
        Last
        <PaginatorIndicator pad />
        <PaginatorIndicator />
      </PaginatorButton>
    </div>
  );
};

export default Paginator;
