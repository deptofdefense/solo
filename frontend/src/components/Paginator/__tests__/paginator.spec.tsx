import React from "react";
import { render, fireEvent } from "test-utils";
import Paginator from "../Paginator";
import { TableInstance } from "react-table";
import { wait } from "@testing-library/react";

describe("Paginator component", () => {
  const nextPageMock = jest.fn();
  const prevPageMock = jest.fn();
  const gotoPageMock = jest.fn();
  const defaultTableInstance = ({
    canPreviousPage: true,
    canNextPage: true,
    previousPage: prevPageMock,
    nextPage: nextPageMock,
    gotoPage: gotoPageMock,
    pageCount: 100,
    state: { pageIndex: 10 }
  } as unknown) as TableInstance;

  afterEach(() => {
    nextPageMock.mockReset();
    prevPageMock.mockReset();
    gotoPageMock.mockReset();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Paginator table={defaultTableInstance} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("goes to first page on first link click", async () => {
    const { getByText } = render(<Paginator table={defaultTableInstance} />);
    const firstPage = getByText("First");
    fireEvent.click(firstPage);
    await wait(() => {
      expect(gotoPageMock).toHaveBeenCalled();
      expect(gotoPageMock.mock.calls[0][0]).toEqual(0);
    });
  });

  it("goes to previous page on previous page click", async () => {
    const { getByText } = render(<Paginator table={defaultTableInstance} />);
    const firstPage = getByText("Previous");
    fireEvent.click(firstPage);
    await wait(() => {
      expect(prevPageMock).toHaveBeenCalled();
    });
  });

  it("goes to next page on next page click", async () => {
    const { getByText } = render(<Paginator table={defaultTableInstance} />);
    const nextPage = getByText("Next");
    fireEvent.click(nextPage);
    await wait(() => {
      expect(nextPageMock).toHaveBeenCalled();
    });
  });

  it("goes to last page on last page click", async () => {
    const { getByText } = render(<Paginator table={defaultTableInstance} />);
    const specificPage = getByText("12");
    fireEvent.click(specificPage);
    await wait(() => {
      expect(gotoPageMock).toHaveBeenCalled();
      expect(gotoPageMock.mock.calls[0][0]).toEqual(11);
    });
  });

  it("goes to the specified page on page link click", async () => {
    const { getByText } = render(<Paginator table={defaultTableInstance} />);
    const lastPage = getByText("Last");
    fireEvent.click(lastPage);
    await wait(() => {
      expect(gotoPageMock).toHaveBeenCalled();
      expect(gotoPageMock.mock.calls[0][0]).toEqual(99);
    });
  });

  it("renders the beginning links properly", async () => {
    const { queryByText } = render(
      <Paginator
        table={{
          ...defaultTableInstance,
          state: {
            ...defaultTableInstance.state,
            pageIndex: 1 // this is page `2` when 1-indexed
          }
        }}
      />
    );
    [1, 2, 3, 4].forEach(num => {
      expect(queryByText(num.toString())).toBeInTheDocument();
    });
    [0, 5].forEach(num => {
      expect(queryByText(num.toString())).not.toBeInTheDocument();
    });
  });

  it("renders the ending links properly", async () => {
    const { queryByText } = render(
      <Paginator
        table={{
          ...defaultTableInstance,
          state: {
            ...defaultTableInstance.state,
            pageIndex: 98 // this is page `99` 1-indexed
          }
        }}
      />
    );
    [97, 98, 99, 100].forEach(num => {
      expect(queryByText(num.toString())).toBeInTheDocument();
    });
    [96, 101].forEach(num => {
      expect(queryByText(num.toString())).not.toBeInTheDocument();
    });
  });

  it("renders the middle links properly", async () => {
    const { queryByText } = render(
      <Paginator
        table={{
          ...defaultTableInstance,
          state: {
            ...defaultTableInstance.state,
            pageIndex: 41 // this is page `42` 1-indexed
          }
        }}
      />
    );
    [40, 41, 42, 43, 44].forEach(num => {
      expect(queryByText(num.toString())).toBeInTheDocument();
    });
    [39, 45].forEach(num => {
      expect(queryByText(num.toString())).not.toBeInTheDocument();
    });
  });
});
