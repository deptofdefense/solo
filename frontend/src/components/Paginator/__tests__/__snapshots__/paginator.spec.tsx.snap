// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-row flex-justify-center margin-top-3"
  >
    <button
      class="usa-button pageLink usa-button--unstyled"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-left fa-w-10 fa-sm "
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-left fa-w-10 fa-sm margin-right-1"
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          fill="currentColor"
        />
      </svg>
      First
    </button>
    <button
      class="usa-button pageLink usa-button--unstyled"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-left fa-w-10 fa-sm margin-right-1"
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          fill="currentColor"
        />
      </svg>
      Previous
    </button>
    <button
      class="usa-button pageLink usa-button--unstyled"
    >
      9
    </button>
    <button
      class="usa-button pageLink usa-button--unstyled"
    >
      10
    </button>
    <button
      class="usa-button pageLink usa-button--outline"
    >
      11
    </button>
    <button
      class="usa-button pageLink usa-button--unstyled"
    >
      12
    </button>
    <button
      class="usa-button pageLink usa-button--unstyled"
    >
      13
    </button>
    <button
      class="usa-button pageLink usa-button--unstyled"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-right fa-w-10 fa-sm margin-left-1"
        data-icon="chevron-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
          fill="currentColor"
        />
      </svg>
    </button>
    <button
      class="usa-button pageLink usa-button--unstyled"
    >
      Last
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-right fa-w-10 fa-sm margin-left-1"
        data-icon="chevron-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-right fa-w-10 fa-sm "
        data-icon="chevron-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
